### ISSUE TITLE
Operation timed out when using where on a set of 100.000 documents

### ISSUE THREAD

**Issue by mvandenbrink** [stargate/stargate#905]
Used the latest Stargate version 1.0.22 to test the new where functionality. My initial set had errors after the upgrade to 1.0.22, so I started with a new empty Stargate and Cassandra setup in Docker. With an initial load of 10.000 documents, I got results from the HTTP GET and with query param ?where=. But on a larger set with 100.000 documents loaded, the HTTP Get works, but if I use the query param ?where= I got a read time out HTTP 500 error.

Stargate log:

```
ERROR [jersey-server-managed-async-executor-15] 2021-04-28 18:30:11,632 DocumentResourceV2.java:844 - Error when executing request
java.util.concurrent.CompletionException: org.apache.cassandra.stargate.exceptions.ReadTimeoutException: Operation timed out - received only 0 responses.
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:661)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
        at io.stargate.db.cassandra.impl.CassandraPersistence.lambda$runOnExecutor$1(CassandraPersistence.java:281)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:165)
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$LocalSessionFutureTask.run(AbstractLocalAwareExecutorService.java:137)
        at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:113)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.cassandra.stargate.exceptions.ReadTimeoutException: Operation timed out - received only 0 responses.
        at io.stargate.db.cassandra.impl.Conversion.toExternal(Conversion.java:307)
        at io.stargate.db.cassandra.impl.Conversion.convertInternalException(Conversion.java:475)
        at io.stargate.db.cassandra.impl.CassandraPersistence$CassandraConnection.lambda$executeRequestOnExecutor$0(CassandraPersistence.java:462)
        at io.stargate.db.cassandra.impl.CassandraPersistence.lambda$runOnExecutor$1(CassandraPersistence.java:272)
        ... 5 common frames omitted
        Suppressed: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 0 responses.
                at org.apache.cassandra.service.ReadCallback.awaitResults(ReadCallback.java:138)
                at org.apache.cassandra.service.ReadCallback.get(ReadCallback.java:143)
                at org.apache.cassandra.service.StorageProxy$SingleRangeResponse.waitForResponse(StorageProxy.java:2071)
                at org.apache.cassandra.service.StorageProxy$SingleRangeResponse.computeNext(StorageProxy.java:2081)
                at org.apache.cassandra.service.StorageProxy$SingleRangeResponse.computeNext(StorageProxy.java:2054)
                at org.apache.cassandra.utils.AbstractIterator.hasNext(AbstractIterator.java:47)
                at org.apache.cassandra.db.transform.BaseIterator.tryGetMoreContents(BaseIterator.java:144)
                at org.apache.cassandra.db.transform.BaseIterator.hasMoreContents(BaseIterator.java:111)
                at org.apache.cassandra.db.transform.BasePartitions.hasNext(BasePartitions.java:105)
                at org.apache.cassandra.service.StorageProxy$RangeCommandIterator.computeNext(StorageProxy.java:2129)
                at org.apache.cassandra.service.StorageProxy$RangeCommandIterator.computeNext(StorageProxy.java:2092)
                at org.apache.cassandra.utils.AbstractIterator.hasNext(AbstractIterator.java:47)
                at org.apache.cassandra.db.transform.BasePartitions.hasNext(BasePartitions.java:92)
                at org.apache.cassandra.cql3.statements.SelectStatement.process(SelectStatement.java:786)
                at org.apache.cassandra.cql3.statements.SelectStatement.processResults(SelectStatement.java:438)
                at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:416)
                at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:289)
                at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:117)
                at org.apache.cassandra.cql3.QueryProcessor.processStatement(QueryProcessor.java:225)
                at org.apache.cassandra.cql3.QueryProcessor.processPrepared(QueryProcessor.java:532)
                at org.apache.cassandra.cql3.QueryProcessor.processPrepared(QueryProcessor.java:509)
                at io.stargate.db.cassandra.impl.StargateQueryHandler.processPrepared(StargateQueryHandler.java:200)
                at org.apache.cassandra.transport.messages.ExecuteMessage.execute(ExecuteMessage.java:146)
                at io.stargate.db.cassandra.impl.CassandraPersistence$CassandraConnection.lambda$executeRequestOnExecutor$0(CassandraPersistence.java:455)
                ... 6 common frames omitted
INFO  [jersey-server-managed-async-executor-15] 2021-04-28 18:30:11,669 Slf4jRequestLogWriter.java:62 - 172.23.0.1 - - [28/Apr/2021:18:30:01 +0000] "GET /v2/namespaces/test/collections/testset HTTP/1.1" 500 154 "-" "PostmanRuntime/7.28.0" 10072
```

Steps that I took:

```
1. create collection
2. Load 10.000 documents
  81593 ms, avg 8.16ms per inserted document
3. GET ?page-size=20
  HTTP 200, 110 ms
4. drop collection
5. create collection
6. Load 100.000 documents
  813395 ms, avg 8.13ms per inserted document
7. GET ?page-size=20
  HTTP 200, 26 ms
8. GET ?where={ "plaats": {"$eq": "Leeuwarden" } }&page-size=20
  HTTP 500, 10.08 s
{
    "description": "Server error: org.apache.cassandra.stargate.exceptions.ReadTimeoutException: Operation timed out - received only 0 responses.",
    "code": 500
}
```

**Comment by dimas-b** [comment]
@mvandenbrink : Thanks for filing this issue.

Would you be able to provide an end-to-end test script for you use case (using fake data for testing, or course)?

If that is not possible, could you provide some rough details about the structure and complexity of your documents?

Thanks!

FYI: we're actively working on query improvements under #828

**Comment by mvandenbrink** [comment]
Hi @dimas-b,

I use real live data, so I can't provide a script with fake data. I have translated the document structure that is used. Hope this can help.

```
{
    place: "name of city",
    isMainBranchOffice: true,
    legalForm: {
      code: "99",
      description: "description string",
    },
    activity: [
      {
        code: "9999",
        description: "description string",
        isMainActivity: true
      },
      {
        code: "9999",
        description: "description string",
        isMainActivity: false
     }
    ],
    import: false,
    export: false,
    dateEstablishment: 20200101,
    dateEstablishmentDocument: 20200101,
    dateRegistration: 20200101,
    dateBranchOffice: 20200101,
    dateBranchOfficeAddress: 20200101,
    dateDeregistration: 20200101,
    dateContinuation: 20200101,
    surseance: {
      indicator: "X",
      description: "description string",
    },
    financialYear: 2020,
    dateDepositAnnualReport: 20200101,
    active: true,
}
```

**Comment by ivansenic** [comment]
@mvandenbrink Thanks for sharing.. Can you maybe app. say how many docs have the `place` you are looking for using the `$eq` filter?

**Comment by mvandenbrink** [comment]
All the documents have a `place`. From the 100.000 documents I counted 380 documents with `place: "Leeuwarden"`.

**Comment by dimas-b** [comment]
@mvandenbrink : Could you also post the Stargate query URL you use (real host name is not required, of course)? Thx!

**Comment by mvandenbrink** [comment]
Of course: `{{baseUrl}}/v2/namespaces/test/collections/testset?page-size=20&where={ "plaats": {"$eq": "Leeuwarden" } }`

**Comment by EricBorczuk** [comment]
Hi @mvandenbrink,

We made a few changes to querying recently - but I wouldn't think that it would affect performance for a `where` that is simple like yours. I'm going to try to investigate this by reproducing locally.

Sorry about the possible regression here, we're making some big code changes to, among other things, support `OR` queries in the where clause; I did test to see that there wasn't a performance regression before merging the change though, so I'm definitely going to investigate this one :)

**Comment by EricBorczuk** [comment]
Hi Marcel,

I investigated the issue you were seeing, and I found something that may have been causing (at the very least it was exacerbating) your issue.

Using the load script with fake data that you sent me a while back, I loaded a local database with 100000 documents. Although the fake documents from a while back are quite a bit smaller, I was able to see a surprisingly high latency (not seconds, but hundreds of milliseconds). When I loaded 300k documents instead, that latency bumped up to nearly a second per request locally.

After looking into this, I realized that in implementing the new logic for filtering, I reversed two important clauses which caused documents that weren't going to be returned to be fetched from the database.  It makes sense that my performance tests didn't catch this because they are "needle in a haystack" tests; I was loading 10^5 documents and then filtering through them where only a very small number matched.

At any rate, this inefficiency is fixed in https://github.com/stargate/stargate/pull/914

That PR makes response time 4x faster for 100k of these smaller documents, so I would imagine a similar-or-greater effect for your database.
